### PR TITLE
Drop experimental reuse fragments optimization from RS QP

### PR DESCRIPTION
**Pull request by dariuszkuc**
Drop support for the experimental reuse fragment query optimization from Rust QP. This implementation was not only very slow but also very buggy due to its complexity. `experimental_reuse_query_fragments` option can be applied for the JS QP only.

Auto generation of fragments is a much simpler (and faster) algorithm that in most cases produces better results. Fragment auto generation is the default optimization since v1.58 release.

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

**Comment by goto-bus-stop** [comment]
We should also make sure that people aren't opting back into the old algorithm en masse in 1.58.0. But this is a very exciting diff 😁

**Review by lrlna — CHANGES_REQUESTED** [review]
This PR also needs to update [docs](https://github.com/apollographql/router/blob/dev/docs/source/reference/router/configuration.mdx#fragment-generation-and-reuse) for fragment generation and reuse.

Otherwise really exciting to get rid of all the dead code!!

**Review by lrlna — CHANGES_REQUESTED** [review]
I am so sorry, I read a tracing::warn message you wrote and realised this is going to need a [config migration](https://github.com/apollographql/router/blob/dev/apollo-router/src/configuration/migrations/README.md). For the time being it can be a `log`, and for the next release we can do a `delete`.

File: apollo-router/src/configuration/snapshots/apollo_router__configuration__tests__schema_generation.snap [hunk 1]
```diff
@@ -6644,7 +6644,7 @@ expression: "&schema"
         },
         "experimental_reuse_query_fragments": {
           "default": null,
-          "description": "Enable reuse of query fragments Default: depends on the federation version",
+          "description": "Enable reuse of query fragments\n\nThis feature is deprecated and will be removed in next release. It is only available in JS QP. Use generate_query_fragments instead.",
```
Comment (dariuszkuc) [inline apollo-router/src/configuration/snapshots/apollo_router__configuration__tests__schema_generation.snap:6647]:
```suggestion
          "description": "Enable reuse of query fragments\n\nThis feature is deprecated and will be removed in next release.\nThe config can only be set when the legacy query planner is explicitly enabled.",
```

**Review by lrlna — APPROVED** [review]
Thank you!

**Pull request merged**